File: lib/services/device_service.dart
import 'package:device_info_plus/device_info_plus.dart';
import 'package:flutter/foundation.dart';
import '../models/device_binding.dart';
import 'supabase_service.dart';

class DeviceService {
  final DeviceInfoPlugin _deviceInfo = DeviceInfoPlugin();
  final SupabaseService _supabase = SupabaseService();

  /// Get unique device ID
  Future<String> getDeviceId() async {
    try {
      if (kIsWeb) {
        final webInfo = await _deviceInfo.webBrowserInfo;
        return webInfo.userAgent ?? 'web-client';
      } else if (defaultTargetPlatform == TargetPlatform.android) {
        final androidInfo = await _deviceInfo.androidInfo;
        return androidInfo.id; // Android ID
      } else if (defaultTargetPlatform == TargetPlatform.iOS) {
        final iosInfo = await _deviceInfo.iosInfo;
        return iosInfo.identifierForVendor ?? 'unknown';
      } else {
        return 'unknown';
      }
    } catch (e) {
      debugPrint('Error getting device ID: $e');
      return 'unknown';
    }
  }

  /// Get device information (name, model, etc.)
  Future<Map<String, String>> getDeviceInfo() async {
    try {
      if (kIsWeb) {
        final webInfo = await _deviceInfo.webBrowserInfo;
        return {
          'name': webInfo.browserName.name,
          'model': webInfo.platform ?? 'Web Browser',
          'os': 'Web',
        };
      } else if (defaultTargetPlatform == TargetPlatform.android) {
        final androidInfo = await _deviceInfo.androidInfo;
        return {
          'name': androidInfo.model,
          'model': '${androidInfo.manufacturer} ${androidInfo.model}',
          'os': 'Android ${androidInfo.version.release}',
        };
      } else if (defaultTargetPlatform == TargetPlatform.iOS) {
        final iosInfo = await _deviceInfo.iosInfo;
        return {
          'name': iosInfo.name,
          'model': iosInfo.model,
          'os': 'iOS ${iosInfo.systemVersion}',
        };
      } else {
        return {
          'name': 'Unknown',
          'model': 'Unknown',
          'os': 'Unknown',
        };
      }
    } catch (e) {
      debugPrint('Error getting device info: $e');
      return {
        'name': 'Unknown',
        'model': 'Unknown',
        'os': 'Unknown',
      };
    }
  }

  /// Check if current device is registered for this user
  Future<bool> isDeviceRegistered(String userId) async {
    try {
      final deviceId = await getDeviceId();
      final binding = await _supabase.getDeviceBinding(userId, deviceId);
      return binding != null && binding.isActive;
    } catch (e) {
      debugPrint('Error checking device registration: $e');
      return false;
    }
  }

  /// Register current device for user
  Future<DeviceBinding?> registerDevice(String userId) async {
    try {
      final deviceId = await getDeviceId();
      final deviceInfo = await getDeviceInfo();

      return await _supabase.registerDevice(
        userId: userId,
        deviceId: deviceId,
        deviceName: deviceInfo['name']!,
        deviceModel: deviceInfo['model']!,
      );
    } catch (e) {
      debugPrint('Error registering device: $e');
      return null;
    }
  }

  /// Unregister a device
  Future<bool> unregisterDevice(String deviceId) async {
    try {
      await _supabase.unregisterDevice(deviceId);
      return true;
    } catch (e) {
      debugPrint('Error unregistering device: $e');
      return false;
    }
  }

  /// Update last used timestamp for device
  Future<void> updateDeviceLastUsed(String userId) async {
    try {
      final deviceId = await getDeviceId();
      await _supabase.updateDeviceLastUsed(userId, deviceId);
    } catch (e) {
      debugPrint('Error updating device last used: $e');
    }
  }

  /// Get all registered devices for a user
  Future<List<DeviceBinding>> getUserDevices(String userId) async {
    try {
      return await _supabase.getUserDevices(userId);
    } catch (e) {
      debugPrint('Error getting user devices: $e');
      return [];
    }
  }
}

File: web-next/src/lib/services/deviceService.ts
import { supabase } from '../supabase';

class DeviceService {
    private readonly STORAGE_KEY = 'braand_device_id';

    getDeviceId(): string {
        if (typeof globalThis.window === 'undefined') return 'server';
        let deviceId = localStorage.getItem(this.STORAGE_KEY);
        if (!deviceId) {
            deviceId = `web_${crypto.randomUUID()}`;
            localStorage.setItem(this.STORAGE_KEY, deviceId);
        }
        return deviceId;
    }

    async isDeviceRegistered(userId: string): Promise<boolean> {
        const deviceId = this.getDeviceId();
        const { data, error } = await supabase
            .from('device_bindings')
            .select('id')
            .eq('user_id', userId)
            .eq('device_id', deviceId)
            .eq('is_active', true)
            .maybeSingle();

        if (error) {
            console.error('Error checking device registration:', error);
            return false;
        }

        return !!data;
    }

    async registerDevice(userId: string): Promise<boolean> {
        const deviceId = this.getDeviceId();
        const userAgent = navigator.userAgent;

        // Simple browser identification
        const deviceName = 'Web Browser';
        const deviceModel = userAgent.split(')')[0].split('(')[1] || 'Unknown';

        const { error } = await supabase
            .from('device_bindings')
            .insert({
                user_id: userId,
                device_id: deviceId,
                device_name: deviceName,
                device_model: deviceModel,
                registered_at: new Date().toISOString(),
                is_active: true,
            });

        if (error) {
            console.error('Error registering device:', error);
            return false;
        }

        return true;
    }

    async updateLastUsed(userId: string): Promise<void> {
        const deviceId = this.getDeviceId();
        await supabase
            .from('device_bindings')
            .update({ last_used_at: new Date().toISOString() })
            .eq('user_id', userId)
            .eq('device_id', deviceId);
    }
}

export const deviceService = new DeviceService();
